detectorSegmentations/src/FCCSWHCalPhiRow_k4geo.cpp

```diff
@@ -0,0 +1,674 @@
+  if(m_cellIndexes[layer].size()==0 && m_radii.size() > 0)
+  {
+    double minLayerZ = m_layerEdges[layer].first;
+    double maxLayerZ = m_layerEdges[layer].second;
+
+    // find rows/sequences that fit within the given layer range along z
+    int irow = 0;
+    while( (minLayerZ + (irow+1)* m_dz_row) < (maxLayerZ+0.0001) )
+    {
+      // define the cell index
+      int idx = floor(irow/m_gridSizeRow[layer]) + 1;
+      // add the index if it is not already there
+      if(std::find(m_cellIndexes[layer].begin(),m_cellIndexes[layer].end(),idx) == m_cellIndexes[layer].end()) m_cellIndexes[layer].push_back(idx);
+      irow++;
+    }
+
+    // for the EndCap, do it again but for negative-z part
+    if(m_offsetZ.size() > 1)
+    {
+      irow = 0;
+      while( (minLayerZ + (irow+1) * m_dz_row) < (maxLayerZ+0.0001) )
+      {
+        // define the cell index with negative sign
+        int idx = - ( floor(irow/m_gridSizeRow[layer]) + 1 );
+        // add the index if it is not already there
+        if(std::find(m_cellIndexes[layer].begin(),m_cellIndexes[layer].end(),idx) == m_cellIndexes[layer].end()) m_cellIndexes[layer].push_back(idx);
+        irow++;
+      }
+    }
+
+    // find edges of each cell in the given layer along z axis
+    for(auto idx : m_cellIndexes[layer])
+    {
+      // calculate z-coordinates of the cell edges
+      double z1 = minLayerZ + (idx-1) * m_dz_row * m_gridSizeRow[layer]; // lower edge
+      double z2 = minLayerZ + (idx) * m_dz_row * m_gridSizeRow[layer]; // upper edge
+
+      // for negative-z Endcap, the index is negative (starts from -1!)
+      if(idx < 0)
+      { 
+        z1 = -minLayerZ + (idx) * m_dz_row * m_gridSizeRow[layer]; // lower edge
+        z2 = -minLayerZ + (idx+1) * m_dz_row * m_gridSizeRow[layer]; // upper edge
+      }
+
+      m_cellEdges[layer][idx] = std::make_pair(z1,z2);
+    }
+
+    dd4hep::printout(dd4hep::DEBUG, "FCCSWHCalPhiRow_k4geo","Number of cells in layer %d: %d", layer, m_cellIndexes[layer].size());
+  }
+}
+
+/// create the cell ID based on the position
+CellID FCCSWHCalPhiRow_k4geo::cellID(const Vector3D& /* localPosition */, const Vector3D& globalPosition,
+                          const VolumeID& vID) const {
+
+  // get the row number from volumeID (starts from 0!)
+  int nrow = _decoder->get(vID, "row");
+  // get the layer number from volumeID
+  uint layer = _decoder->get(vID,"layer");
+
+  CellID cID = vID;
+
+  // get the cell index (start from 1!)
+  int idx = floor(nrow/m_gridSizeRow[layer]) + 1;
+
+  // if the hit is in the negative-z part of the Endcap then assign negative index
+  if(m_offsetZ.size() > 1 && globalPosition.z() < 0) idx *= -1;
+
+  _decoder->set(cID, m_rowID, idx);
+  _decoder->set(cID, m_phiID, positionToBin(dd4hep::DDSegmentation::Util::phiFromXYZ(globalPosition), 2 * M_PI / (double)m_phiBins, m_offsetPhi));
+
+  // For endcap, the volume ID comes with "type" field information which would screw up the topo-clustering,
```

Comment:
aren't HCAL barrel and endcap already differentiated by system ID? Is type set to something different from 0 elsewhere?

Reply:
yes, they are already differentiated by systemID.
The "type" is set different from zero in the geo builder:
https://github.com/key4hep/k4geo/blob/main/detector/calorimeter/HCalThreePartsEndcap_o1_v02_geo.cpp#L242

I am not sure if it is used elsewhere, it was already like that and I did not touched it.
I guess this field is set to differentiate different parts of the endcap.